core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+
+import com.linecorp.armeria.common.Flags;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import io.netty.handler.timeout.IdleStateHandler;
+
+/**
+ * This will send {@link Http2PingFrame} when an {@link IdleStateEvent} is emitted by {@link
+ * IdleStateHandler}. Specifically, it will write a PING frame to remote and then expects an ACK
+ * back within configured {@code pingTimeoutMillis}. If no valid response is received in time, then
+ * channel will be closed.
+ *
+ * <p>This class is <b>not</b> thread-safe and all methods are to be called from single thread such
+ * as {@link EventLoop}.
+ */
+@NotThreadSafe
+public class Http2KeepAliveHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(Http2KeepAliveHandler.class);
+
+    private final boolean sendPingsOnNoActiveStreams;
+    private final long pingTimeoutMillis;
+    private final Http2FrameWriter frameWriter;
+    private final ThreadLocalRandom random = ThreadLocalRandom.current();
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
```

Comment:
If `stopwatch` is used only for debug logging, how about assigning null to this field when `logger.isDebugEnabled()` is false?